app/plugin.go

```diff
@@ -400,6 +402,96 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
 	return resp, nil
 }
 
+// GetMarketplacePlugins returns a list of plugins from the marketplace-server,
+// and plugins that are installed locally.
+func (a *App) GetMarketplacePlugins(filter *model.MarketplacePluginFilter) ([]*model.MarketplacePlugin, *model.AppError) {
+	var result []*model.MarketplacePlugin
+	pluginSet := map[string]bool{}
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.config.app_error", nil, "", http.StatusInternalServerError)
+	}
+
+	marketplaceClient, err := marketplace.NewClient(
+		*a.Config().PluginSettings.MarketplaceUrl,
+		a.HTTPService,
+	)
+	if err != nil {
+		return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.marketplace_client.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	marketplacePlugins, err := marketplaceClient.GetPlugins(filter)
+	if err != nil {
+		return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.marketplace_plugins.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	for _, plugin := range marketplacePlugins {
+		installedVersion := ""
+		if pluginsEnvironment != nil {
+			var manifest *model.Manifest
+			if manifest, err = pluginsEnvironment.Manifest(plugin.Manifest.Id); err == nil {
```

Comment:
nit: I'd check `manifest != nil` here. I bet we return an `err` when the manifest doesn't exist, but I'd love to see us fix that in the future, so that `nil, nil` == the equivalent of a 404.

Reply:
see https://github.com/mattermost/mattermost-server/pull/11977#discussion_r321237389